public/app/plugins/panel/datagrid/DataGridPanel.tsx

```diff
@@ -0,0 +1,456 @@
+import { css } from '@emotion/css';
+import DataEditor, {
+  GridCell,
+  Item,
+  GridColumn,
+  GridCellKind,
+  EditableGridCell,
+  GridColumnIcon,
+  GridSelection,
+  CellClickedEventArgs,
+  Rectangle,
+  HeaderClickedEventArgs,
+  SizedGridColumn,
+} from '@glideapps/glide-data-grid';
+import React, { useCallback, useEffect, useState } from 'react';
+
+import {
+  ArrayVector,
+  Field,
+  MutableDataFrame,
+  PanelProps,
+  GrafanaTheme2,
+  getFieldDisplayName,
+  FieldType,
+} from '@grafana/data';
+import { PanelDataErrorView } from '@grafana/runtime';
+// eslint-disable-next-line import/order
+import { useTheme2 } from '@grafana/ui';
+
+import '@glideapps/glide-data-grid/dist/index.css';
+
+import { AddColumn } from './components/AddColumn';
+import { DatagridContextMenu } from './components/DatagridContextMenu';
+import { RenameColumnCell } from './components/RenameColumnCell';
+import { PanelOptions } from './panelcfg.gen';
+import {
+  RenameColumnInputData,
+  clearCellsFromRangeSelection,
+  DatagridContextMenuData,
+  deleteRows,
+  EMPTY_CELL,
+  EMPTY_GRID_SELECTION,
+  getCellWidth,
+  isDatagridEditEnabled,
+  publishSnapshot,
+  RIGHT_ELEMENT_PROPS,
+  TRAILING_ROW_OPTIONS,
+} from './utils';
+
+interface Props extends PanelProps<PanelOptions> {}
+
+export const DataGridPanel: React.FC<Props> = ({ options, data, id, fieldConfig }) => {
+  const [columnWidths, setColumnWidths] = useState<Map<number, number>>(new Map());
+  const [columns, setColumns] = useState<SizedGridColumn[]>([]);
+  const [contextMenuData, setContextMenuData] = useState<DatagridContextMenuData>({ isContextMenuOpen: false });
+  const [renameColumnInputData, setRenameColumnInputData] = useState<RenameColumnInputData>({ isInputOpen: false });
+  const [gridSelection, setGridSelection] = useState<GridSelection>(EMPTY_GRID_SELECTION);
+  const [columnFreezeIndex, setColumnFreezeIndex] = useState<number>(0);
+  const [toggleSearch, setToggleSearch] = useState<boolean>(false);
+  const [isResizeInProgress, setIsResizeInProgress] = useState<boolean>(false);
```

Comment:
I would try to find a way to move state and logic outside the react componet (using useReducer, or StateManagerBase) or external hook. Easier to unit test state and callback logic and makes it easier to reason about the react rendering logic

Reply:
I agree with you on state logic. Testing is underway for this, but I wanted to get eyes on it as soon as possible.